### PR TITLE
relationships: initial implementation

### PR DESCRIPTION
**Pull request by MSusik**
* Introduces Relationship SQLAlchemy model.

* Introduces Node class for the representation of a node in
  relationship graph.

* Includes tests.

Few notes about the module:
+ I'm not sure if I should implement more features (if yes, which ones?).
+ How should I add you (@jirikuncar) as another author of the code? Is second `Signed-off by` enough?

To do:
- [x] - Make everybody discuss this PR.
- [x] - Switch to use `Record`/`JSONAlchemy`/`whatever` instead of using `RecordMetadata` cache.
- [x] - Make tests working.
- [x] - Polish the code.
- [x] - Document `Node` class.
- [x] -Switch from using `JSON` to `JSONB`. -> `sqlalchemy_utils` will probably switch from using `JSON` to `JSONB`.

### CODE REVIEW

File: invenio/modules/relationships/models.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    link_type = db.Column(db.String(255), index=True)
+    model_to = db.Column(db.String(255), index=True)
+    id_to = db.Column(db.Integer(unsigned=True), index=True)
+
+    confidence = db.Column(db.Float, nullable=True)
+
+    def __init__(self, model_from, id_from, link_type, model_to, id_to):
+        """Initialize the relationship between two records.
+
+        Parameters
```
Comment (jirikuncar) [inline invenio/modules/relationships/models.py]: please remove these sections

**Comment by jirikuncar** [comment]
@MSusik what about keeping the original commit and adding your changes further? Otherwise `Co-authored-by: me` would be nice.

File: invenio/modules/relationships/errors.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Errors used by relationship module."""
+
+
+class NoSuchRecordException(Exception):
+
+    """Raised when the record was expected, but the given id is invalid."""
+
+    pass
```
Comment (jirikuncar) [inline invenio/modules/relationships/errors.py]: `pass` is not needed

File: invenio/modules/relationships/models.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            raise WrongModelNameException("Wrong model for record %s" %
+                                          self.id_to)
+        return model
+
+    @utils.session_manager
+    def save(self):
+        db.session.add(self)
+
+
+class Node(object):
```
Comment (jirikuncar) [inline invenio/modules/relationships/models.py]: Please check the original implementation. The idea behind was to allow linking between different models.

File: invenio/modules/relationships/models.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+from .errors import NoSuchRecordException, WrongModelNameException
+
+
+class Relationship(db.Model):
+
+    """Represent a graph edge."""
+
+    uuid = db.Column(UUIDType(binary=False), primary_key=True)
+    model_from = db.Column(db.String(255), index=True)
```
Comment (kaplun) [inline invenio/modules/relationships/models.py]: What is `model_from`/`model_to`? Aren't we going to link records all the time? If so, isn't the model already a property of the record?

File: invenio/modules/relationships/models.py (hunk 1 above)
Comment (kaplun) [inline invenio/modules/relationships/models.py]: So that I can learn: why?

**Comment by kaplun** [comment]
> I'm not sure if I should implement more features (if yes, which ones?).

* We need syntactic sugar in JSONAlchemy to express such the concept of relationship, but it's not yet clear to me how to best express it. @jirikuncar?

* Basically it should be very easy to create in calculated fields in JSONAlchemy such as: record IDs of all the records to that refers to the current one. IDs of all the children record (in case of parent/children relationship), IDs of superseded records, etc.

File: invenio/modules/relationships/models.py (hunk 3 above)
Comment (MSusik) [inline invenio/modules/relationships/models.py]: This allows linking between different models as well. I didn't get the idea with your registry inside `Node` metaclass, though.

File: invenio/modules/relationships/models.py (hunk 4 above)
Comment (MSusik) [inline invenio/modules/relationships/models.py]: I'll explain it in your RFC.

Anyway, it should be easy to query for relationship for a given model. This is the simpliest solution.

**Comment by MSusik** [comment]
How should I replace using `RecordMetadata`?

For example given a test case:
+ Create empty record.
+ Get JSON representation of it (with metadata).
+ Create relationship
+ etc.

It seems that only place where the `JSON` is stored is `RecordMetadata`. So, if I want to get JSON, I believe I should use legacy way to obtain record, create xml, then pass it to `Record.create(master_format="xml" ...)`, and only then use it. Am I correct?

**Comment by MSusik** [comment]
@kaplun:
> We need syntactic sugar

>Basically it should be very easy to create in calculated fields in JSONAlchemy such as

With the methods provided by this PR, it should be very easy to implement both features inside JSONAlchemy.

**Comment by MSusik** [comment]
@jirikuncar Ready to review.

**Comment by MSusik** [comment]
Ready. Not sure why builds don't run correctly.

File: invenio/modules/relationships/api.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Relationship API."""
+
+import importlib
+
+from .config import RELATIONSHIP_ENGINE, RELATIONSHIP_NODE
+
+try:
+    __engine = __import__('flask').current_app.config['RELATIONSHIP_ENGINE']
```
Comment (jirikuncar) [inline invenio/modules/relationships/api.py]: please use `LocalProxy` instead and register the engine and node on application context

File: invenio/modules/relationships/engines/sqlalchemy.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        :param outwards: boolean
+            Get the vertices that are destinations from this node.
+        :param link_type: string
+            Type of the edges. Optional.
+
+        :returns: set
+            Nodes that are neighbours of this node.
+        """
+        edges = self.edges(inwards, outwards, link_type, loops_doubled=False)
+        set1 = {SQLAlchemyNode(id=node.id_from,
```
Comment (jirikuncar) [inline invenio/modules/relationships/engines/sqlalchemy.py]: is it Python 2.6 friendly? we should check that nobody is planning to use *Invenio 2.0* on machines with Python < 2.7

File: invenio/modules/relationships/node.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+    __storage_engine_registry__ = []
+    __entity_types_registry__ = {}
+
+    def __init__(cls, name, bases, dct):
+        """Register cls to type registry."""
+        if hasattr(cls, '__storagename__'):
+            cls.__storage_engine_registry__.append(cls)
+            if cls.__storagename__ not in cls.__entity_types_registry__:
+                print "UPDATER"
```
Comment (jirikuncar) [inline invenio/modules/relationships/node.py]: forgotten print statements

File: invenio/modules/relationships/api.py (hunk 5 above)
Comment (MSusik) [inline invenio/modules/relationships/api.py]: There is a small problem with a `LocalProxy`. As `SmartJSON` inherits from `Node`, when the registries are run, before the application context is ready, the name is imported from this file. If I use `lazy_import` to do it, it doesn't change anything, as it still has to be imported when `SmartJSON` is defined.

To be honest, I can't see a nice solution to this problem. The one in this file is definitely wrong, although when sb will implement another engine, he will be forced to improve it.

As I mentioned before: I don't know what are the plans for integrating new `JSONAlchemy` package inside `Invenio`. As the new module is dependant on it, probably the implementation should be changed.

**Comment by jirikuncar** [comment]
Superseded by #2823.

**Pull request closed without merging**